### PR TITLE
Fix ICE building gluon_vm

### PR DESCRIPTION
**Pull request by nikomatsakis**
The problem was due to various places we were failing to propagate obligations. I think I got them mostly correct, but I didn't get around to writing test cases for each case.

r? @eddyb or @arielb1

### CODE REVIEW

File: src/librustc_typeck/check/compare_method.rs [hunk 1]
```diff
@@ -301,10 +301,10 @@ fn compare_predicate_entailment<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
         debug!("compare_impl_method: trait_fty={:?}", trait_fty);
 
         let sub_result = infcx.sub_types(false, &cause, impl_fty, trait_fty)
-            .map(|InferOk { obligations, .. }| {
-                // FIXME(#32730) propagate obligations
-                assert!(obligations.is_empty());
-            });
+                              .map(|InferOk { obligations, .. }| {
+                                  fulfillment_cx.borrow_mut()
+                                                .register_predicate_obligations(infcx, obligations);
```
Comment (eddyb) [inline src/librustc_typeck/check/compare_method.rs]: IMO this should be using the equivalent `inh` instead... or not construct an `Inherited` at all.

**Review by eddyb — APPROVED** [review]
r=me regardless of what you choose to do with `Inherited`

**Comment by eddyb** [comment]
Wait, #41325 isn't merged yet, that explains why that code looks old. What are the intended interactions between the two PRSs?

**Comment by nikomatsakis** [comment]
Hmm, doesn't seem that they overlap much, but I can try to merge on top of that one.

**Comment by eddyb** [comment]
#41325 merged, please rebase.

**Pull request merged**